File: day7.pl
%% -*- prolog -*-

:- use_module(library(dcg/basics)).

lines --> [].
lines --> line, lines.
line --> string(Ps), " (", { atom_codes(P, Ps) }, integer(V), ")", { asserta(value(P, V)) },
         ("\n" ; " -> ", children(Cs), { forall(member(C, Cs), asserta(parent(P,C))) }).
children([C]) --> child(C), "\n".
children([C|Cs]) --> child(C), ", ", children(Cs).
child(C) --> string_without(",\n", Cs), { atom_codes(C, Cs) }.

part_1(N) :- value(N, _), not(parent(_, N)).

kids_and_weights(N, Cs, Ws) :- findall(C, parent(N, C), Cs), maplist(weight, Cs, Ws).

weight(N, W) :- kids_and_weights(N, _, Ws), sumlist(Ws, Wc), value(N, Wu), W is Wu + Wc.

outsider(L, Crowd, Outsider) :- outsider(L, _, _, Crowd, Outsider), !.
outsider([], A, B, C, O) :- A \= B, C = A, O = B.
outsider([A|[B|Rest]], A, B, C, O) :- outsider(Rest, A, B, C, O).
outsider([A|Rest], A, B, C, O) :- outsider(Rest, A, B, C, O).
outsider([B|Rest], A, B, C, O) :- outsider(Rest, B, A, C, O).

unbalanced(N, U) :-
    kids_and_weights(N, Cs, Ws),
    outsider(Ws, _, O) -> nth1(I, Ws, O), nth1(I, Cs, U1), unbalanced(U1, U); N = U.

should_be(N, DesiredWeight) :-
    parent(P, N), kids_and_weights(P, _, Ws), outsider(Ws, C, O), value(N, V),
    DesiredWeight is V-(O-C).

main :-
    once(phrase_from_file(lines, "day7.in")),
    part_1(Root),
    unbalanced(Root, Unbalanced), should_be(Unbalanced, DesiredWeight),
    format('part 1: ~w~npart 2: ~w should be ~w~n', [Root, Unbalanced, DesiredWeight]).

File: day7.py
from collections import defaultdict

parent,node,child = {},{},defaultdict(list)

def process(line):
    node[line[0]] = int(line[1][1:-1])
    for n in line[3:]:
        if n[-1] == ',': n = n[:-1]
        assert n not in parent
        parent[n] = line[0]
        child[line[0]].append(n)

with open('day7.in') as f:
    for line in f: process(line.split())

n = list(parent.keys())[0]
while n in parent: n = parent[n]
print(n)

def compute(n):
    return node[n] + sum(compute(c) for c in child[n])

def weights_of(root):
    w = node[root]
    weights = []
    for c in child[root]:
        weights.append((c, node[c], compute(c)))
    return weights

def find_unbalanced(n):
    weights = weights_of(n)
    if len(weights) == 0: return None
    base = weights[0][2]
    for w in weights:
        (node,weight,total) = w
        if total != base:
            x = find_unbalanced(node)
            if x is None: return (w,weight-(total-base))
            return x
    return None

print(find_unbalanced(n))
